src/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.HandleNull.cs

```diff
@@ -25,6 +25,12 @@ private static bool HandleNull(ref Utf8JsonReader reader, ref ReadStack state, J
             }
 
             JsonPropertyInfo propertyInfo = state.Current.JsonPropertyInfo;
+
+            if (!propertyInfo.ShouldDeserialize)
+            {
+                return false;
```

Comment:
If I understood well we should return true only if we're sure to be at the end of json object root